File: collector/src/main/java/com/bitmovin/analytics/stateMachines/PlayerStateMachine.java
package com.bitmovin.analytics.stateMachines;

import android.os.CountDownTimer;
import android.os.Handler;
import android.util.Log;
import com.bitmovin.analytics.BitmovinAnalytics;
import com.bitmovin.analytics.BitmovinAnalyticsConfig;
import com.bitmovin.analytics.CustomDataHelpers;
import com.bitmovin.analytics.data.CustomData;
import com.bitmovin.analytics.data.ErrorCode;
import com.bitmovin.analytics.data.ErrorData;
import com.bitmovin.analytics.enums.AnalyticsErrorCodes;
import com.bitmovin.analytics.enums.VideoStartFailedReason;
import com.bitmovin.analytics.utils.Util;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class PlayerStateMachine {
    private static final String TAG = "PlayerStateMachine";
    private final BitmovinAnalyticsConfig config;
    private List<StateMachineListener> listeners = new ArrayList<StateMachineListener>();
    private PlayerState<?> currentState;
    private long elapsedTimeOnEnter = 0;
    private long startupTime = 0;
    // Setting a playerStartupTime of 1 to workaround dashboard issue (only for the
    // first startup sample, in case the collector supports multiple sources)
    private long playerStartupTime = 1L;
    private boolean startupFinished = false;
    private long elapsedTimeSeekStart = 0;
    private long videoTimeStart;
    private long videoTimeEnd;
    private String impressionId;
    private Handler heartbeatHandler = new Handler();
    private int currentRebufferingIntervalIndex = 0;
    private static List<Integer> rebufferingIntervals =
            Arrays.asList(3000, 5000, 10000, 30000, 59700);
    private int heartbeatDelay = 59700; // default to 60 seconds
    private final BitmovinAnalytics analytics;
    private VideoStartFailedReason videoStartFailedReason;
    private int qualityChangeCount = 0;
    protected boolean isQualityChangeTimerRunning = false;

    public PlayerStateMachine(BitmovinAnalyticsConfig config, BitmovinAnalytics analytics) {
        this.config = config;
        this.analytics = analytics;
        this.heartbeatDelay = this.config.getHeartbeatInterval();
        resetStateMachine();
    }

    public void enableHeartbeat() {
        heartbeatHandler.postDelayed(
                new Runnable() {
                    public void run() {
                        triggerHeartbeat();
                        heartbeatHandler.postDelayed(this, heartbeatDelay);
                    }
                },
                heartbeatDelay);
    }

    public void disableHeartbeat() {
        heartbeatHandler.removeCallbacksAndMessages(null);
    }

    public void enableRebufferHeartbeat() {
        heartbeatHandler.postDelayed(
                new Runnable() {
                    public void run() {
                        triggerHeartbeat();
                        currentRebufferingIntervalIndex =
                                Math.min(
                                        currentRebufferingIntervalIndex + 1,
                                        rebufferingIntervals.size() - 1);
                        heartbeatHandler.postDelayed(
                                this, rebufferingIntervals.get(currentRebufferingIntervalIndex));
                    }
                },
                rebufferingIntervals.get(currentRebufferingIntervalIndex));
    }

    public void disableRebufferHeartbeat() {
        currentRebufferingIntervalIndex = 0;
        heartbeatHandler.removeCallbacksAndMessages(null);
    }

    private void triggerHeartbeat() {
        long elapsedTime = Util.getElapsedTime();
        videoTimeEnd = analytics.getPosition();
        for (StateMachineListener listener : getListeners()) {
            listener.onHeartbeat(elapsedTime - elapsedTimeOnEnter);
        }
        elapsedTimeOnEnter = elapsedTime;
        videoTimeStart = videoTimeEnd;
    }

    private void resetSourceRelatedState() {
        disableHeartbeat();
        disableRebufferHeartbeat();
        this.impressionId = Util.getUUID();
        this.videoStartFailedReason = null;
        startupTime = 0;
        startupFinished = false;
        videoStartTimeout.cancel();
        qualityChangeResetTimeout.cancel();
        rebufferingTimeout.cancel();
        resetQualityChangeCount();
        analytics.resetSourceRelatedState();
    }

    public void resetStateMachine() {
        resetSourceRelatedState();
        setCurrentState(PlayerStates.READY);
    }

    public void sourceChange(long oldVideoTime, long newVideoTime, boolean shouldStartup) {
        transitionState(PlayerStates.SOURCE_CHANGED, oldVideoTime, null);
        resetSourceRelatedState();

        if (shouldStartup) {
            transitionState(PlayerStates.STARTUP, newVideoTime, null);
        }
    }

    public synchronized <T> void transitionState(
            PlayerState<T> destinationPlayerState, long videoTime) {
        transitionState(destinationPlayerState, videoTime, null);
    }

    public synchronized <T> void transitionState(
            PlayerState<T> destinationPlayerState, long videoTime, T data) {
        if (!this.isTransitionAllowed(currentState, destinationPlayerState)) {
            return;
        }

        long elapsedTime = Util.getElapsedTime();
        videoTimeEnd = videoTime;

        Log.d(
                TAG,
                "Transitioning from "
                        + currentState.toString()
                        + " to "
                        + destinationPlayerState.toString());

        currentState.onExitState(this, elapsedTime, destinationPlayerState);
        this.elapsedTimeOnEnter = elapsedTime;
        videoTimeStart = videoTimeEnd;
        destinationPlayerState.onEnterState(this, data);
        setCurrentState(destinationPlayerState);
    }

    private boolean isTransitionAllowed(PlayerState<?> currentState, PlayerState<?> destination) {
        if (destination == this.currentState) {
            return false;
        } else if (this.currentState == PlayerStates.EXITBEFOREVIDEOSTART) {
            return false;
        }
        // no state transitions like PLAYING or PAUSE during AD
        else if (currentState == PlayerStates.AD
                && (destination != PlayerStates.ERROR && destination != PlayerStates.ADFINISHED)) {
            return false;
        } else if (currentState == PlayerStates.READY
                && (destination != PlayerStates.ERROR
                        && destination != PlayerStates.EXITBEFOREVIDEOSTART
                        && destination != PlayerStates.STARTUP
                        && destination != PlayerStates.AD)) {
            return false;
        }

        return true;
    }

    public void error(long videoTime, ErrorCode errorCode) {
        transitionState(PlayerStates.ERROR,videoTime, errorCode);
    }

    public boolean isStartupFinished() {
        return startupFinished;
    }

    public void setStartupFinished(boolean startupFinished) {
        this.startupFinished = startupFinished;
    }

    public void addListener(StateMachineListener toAdd) {
        listeners.add(toAdd);
    }

    public void removeListener(StateMachineListener toRemove) {
        listeners.remove(toRemove);
    }

    List<StateMachineListener> getListeners() {
        return listeners;
    }

    public PlayerState<?> getCurrentState() {
        return currentState;
    }

    private void setCurrentState(final PlayerState<?> newPlayerState) {
        this.currentState = newPlayerState;
    }

    public long getStartupTime() {
        return startupTime;
    }

    public void addStartupTime(long elapsedTime) {
        this.startupTime += elapsedTime;
    }

    public String getImpressionId() {
        return impressionId;
    }

    public long getVideoTimeStart() {
        return videoTimeStart;
    }

    public long getVideoTimeEnd() {
        return videoTimeEnd;
    }

    public long getElapsedTimeOnEnter() {
        return elapsedTimeOnEnter;
    }

    public long getElapsedTimeSeekStart() {
        return elapsedTimeSeekStart;
    }

    public long getAndResetPlayerStartupTime() {
        long playerStartupTime = this.playerStartupTime;
        this.playerStartupTime = 0;
        return playerStartupTime;
    }

    public void setElapsedTimeSeekStart(long elapsedTimeSeekStart) {
        this.elapsedTimeSeekStart = elapsedTimeSeekStart;
    }

    public VideoStartFailedReason getVideoStartFailedReason() {
        return videoStartFailedReason;
    }

    public void setVideoStartFailedReason(VideoStartFailedReason videoStartFailedReason) {
        this.videoStartFailedReason = videoStartFailedReason;
    }

    protected CountDownTimer videoStartTimeout =
            new CountDownTimer(Util.VIDEOSTART_TIMEOUT, 1000) {
                @Override
                public void onTick(long millisUntilFinished) {}

                @Override
                public void onFinish() {
                    Log.d(TAG, "VideoStartTimeout finish");
                    setVideoStartFailedReason(VideoStartFailedReason.TIMEOUT);
                    transitionState(PlayerStates.EXITBEFOREVIDEOSTART, 0, null);
                }
            };

    public void pause(long position) {
        if (isStartupFinished()) {
            transitionState(PlayerStates.PAUSE, position);
        } else {
            transitionState(PlayerStates.READY, position);
        }
    }

    public void startAd(long position) {
        transitionState(PlayerStates.AD, position);
        startupTime = 0;
    }

    public boolean isQualityChangeEventEnabled() {
        return this.qualityChangeCount <= Util.ANALYTICS_QUALITY_CHANGE_COUNT_THRESHOLD;
    }

    public void increaseQualityChangeCount() {
        this.qualityChangeCount++;
    }

    protected void resetQualityChangeCount() {
        this.qualityChangeCount = 0;
    }

    public void changeCustomData(
            long position, CustomData customData, CustomDataHelpers.Setter customDataSetter) {
        PlayerState originalState = this.getCurrentState();
        boolean shouldTransition =
                originalState == PlayerStates.PLAYING || originalState == PlayerStates.PAUSE;
        if (shouldTransition) {
            this.transitionState(PlayerStates.CUSTOMDATACHANGE, position);
        }
        customDataSetter.setCustomData(customData);
        if (shouldTransition) {
            this.transitionState(originalState, position);
        }
    }

    protected CountDownTimer qualityChangeResetTimeout =
            new CountDownTimer(Util.ANALYTICS_QUALITY_CHANGE_COUNT_RESET_INTERVAL, 1000) {

                @Override
                public void onTick(long millisUntilFinished) {
                    isQualityChangeTimerRunning = true;
                }

                @Override
                public void onFinish() {
                    Log.d(TAG, "qualityChangeResetTimeout finish");
                    resetQualityChangeCount();
                    isQualityChangeTimerRunning = false;
                }
            };

    protected CountDownTimer rebufferingTimeout =
            new CountDownTimer(Util.REBUFFERING_TIMEOUT, 1000) {
                @Override
                public void onTick(long millisUntilFinished) {}

                @Override
                public void onFinish() {
                    Log.d(TAG, "rebufferingTimeout finish");
                    error(analytics.getPosition(),
                            AnalyticsErrorCodes.ANALYTICS_BUFFERING_TIMEOUT_REACHED.getErrorCode());
                    disableRebufferHeartbeat();
                    resetStateMachine();
                }
            };
}

File: collector/src/main/java/com/bitmovin/analytics/stateMachines/PlayerStateMachine.kt
package com.bitmovin.analytics.stateMachines

import android.os.Handler
import android.os.Looper
import com.bitmovin.analytics.BitmovinAnalytics
import com.bitmovin.analytics.ObservableSupport
import com.bitmovin.analytics.adapters.PlayerContext
import com.bitmovin.analytics.data.DeviceInformationProvider
import com.bitmovin.analytics.dtos.ErrorCode
import com.bitmovin.analytics.dtos.SubtitleDto
import com.bitmovin.analytics.enums.AnalyticsErrorCodes
import com.bitmovin.analytics.enums.VideoStartFailedReason
import com.bitmovin.analytics.utils.BitmovinLog
import com.bitmovin.analytics.utils.Util
import com.bitmovin.analytics.utils.Util.HEARTBEAT_INTERVAL

class PlayerStateMachine(
    private val analytics: BitmovinAnalytics,
    internal val bufferingTimeoutTimer: ObservableTimer,
    internal val qualityChangeEventLimiter: QualityChangeEventLimiter,
    internal val videoStartTimeoutTimer: ObservableTimer,
    private val playerContext: PlayerContext,
    looper: Looper,
    private val deviceInformationProvider: DeviceInformationProvider,
    private val heartbeatHandler: Handler = Handler(looper),
) {
    internal val listeners = ObservableSupport<StateMachineListener>()

    var currentState: PlayerState<*> = PlayerStates.READY
        private set
    var startupTime: Long = 0
        private set

    private var elapsedTimeOnEnter: Long = 0

    var isStartupFinished = false

    var videoTimeStart: Long = 0
        private set
    var videoTimeEnd: Long = 0
        private set

    private var currentRebufferingIntervalIndex = 0
    private val heartbeatDelay = HEARTBEAT_INTERVAL.toLong() // 60 seconds
    var videoStartFailedReason: VideoStartFailedReason? = null

    init {
        bufferingTimeoutTimer.subscribe(::onRebufferingTimerFinished)
        videoStartTimeoutTimer.subscribe(::onVideoStartTimeoutTimerFinished)
        resetStateMachine()
    }

    fun isInStartupState(): Boolean {
        return currentState === PlayerStates.STARTUP
    }

    fun enableHeartbeat() {
        heartbeatHandler.postDelayed(
            object : Runnable {
                override fun run() {
                    val continueHeartbeat = checkAndTriggerPlayingSample()
                    if (continueHeartbeat) {
                        heartbeatHandler.postDelayed(this, heartbeatDelay)
                    }
                }
            },
            heartbeatDelay,
        )
    }

    fun disableHeartbeat() {
        heartbeatHandler.removeCallbacksAndMessages(null)
    }

    fun enableRebufferHeartbeat() {
        heartbeatHandler.postDelayed(
            object : Runnable {
                override fun run() {
                    triggerSample()
                    currentRebufferingIntervalIndex =
                        Math.min(
                            currentRebufferingIntervalIndex + 1,
                            rebufferingIntervals.size - 1,
                        )
                    heartbeatHandler.postDelayed(
                        this,
                        rebufferingIntervals[currentRebufferingIntervalIndex].toLong(),
                    )
                }
            },
            rebufferingIntervals[currentRebufferingIntervalIndex].toLong(),
        )
    }

    fun disableRebufferHeartbeat() {
        currentRebufferingIntervalIndex = 0
        heartbeatHandler.removeCallbacksAndMessages(null)
    }

    // last sample of session when collector is detached
    // player destroyed, or source unloaded
    fun triggerLastSampleOfSession() {
        // we only send the last sample if we are in a state that is not triggering
        // an immediate sample by itself and ignore PAUSE, since that one is not relevant as a last sample
        if (currentState === PlayerStates.PLAYING ||
            currentState === PlayerStates.BUFFERING ||
            currentState === PlayerStates.AD
        ) {
            triggerSample()
        }
    }

    // Exoplayer and IVsPlayer do not have a 'playerWasReleased' event, so we can not detect when
    // the player was released. This is problematic when a customer releases player but does not
    // detach our collectors, as we do not transition into pause state and will continue sending
    // samples. The below check prevents this from happening.
    // Further, on FireOS 8 we have seen that the player is in playing state but not actually
    // playing, which also leads to heartbeat samples for long periods of time, which mingles with our data.
    // This is true for media3 and bitmovin player (as of 2025-05-14)
    // -> https://bitmovin.atlassian.net/browse/AN-4679
    private fun checkAndTriggerPlayingSample(): Boolean {
        if (playerContext.isPlaying()) {
            // workaround for fireos 8 since the playing state
            // is not reliable when the display is turned off
            // https://bitmovin.atlassian.net/browse/AN-4679
            if (deviceInformationProvider.isFireOs8OrHigher) {
                if (playerIsMakingProgressBetweenHeartbeats()) {
                    triggerSample()
                    return true
                } else {
                    pause(playerContext.position)
                    return false
                }
            } else {
                triggerSample()
                return true
            }
        } else {
            // transition into pause state when statemachine is in PLAYING state but player not actually playing
            pause(playerContext.position)
            return false
        }
    }

    private var lastPlayerPosition: Long = 0

    // we verify if the player is changing the position since the last heartbeat
    private fun playerIsMakingProgressBetweenHeartbeats(): Boolean {
        val currentPosition = playerContext.position

        try {
            return currentPosition != lastPlayerPosition
        } finally {
            lastPlayerPosition = currentPosition
        }
    }

    // Trigger sample that is not caused by a player event directly (heartbeat, ssai ad block, detaching,...
    private fun triggerSample(ssaiRelated: Boolean = false) {
        val elapsedTime = Util.elapsedTime
        videoTimeEnd = playerContext.position
        listeners.notify { it.onTriggerSample(this, elapsedTime - elapsedTimeOnEnter, ssaiRelated) }
        elapsedTimeOnEnter = elapsedTime
        videoTimeStart = videoTimeEnd
    }

    private fun resetSourceRelatedState() {
        disableHeartbeat()
        disableRebufferHeartbeat()
        videoStartTimeoutTimer.cancel()
        bufferingTimeoutTimer.cancel()
        qualityChangeEventLimiter.reset()
        analytics.resetSourceRelatedState()

        videoStartFailedReason = null
        isStartupFinished = false
        startupTime = 0
    }

    fun resetStateMachine() {
        resetSourceRelatedState()
        currentState = PlayerStates.READY
    }

    @Synchronized
    fun <T> transitionState(
        destinationPlayerState: PlayerState<T>,
        videoTime: Long,
    ) {
        transitionState(destinationPlayerState, videoTime, null)
    }

    @Synchronized
    fun <T> transitionState(
        destinationPlayerState: PlayerState<T>,
        videoTime: Long,
        data: T?,
    ) {
        if (!isTransitionAllowed(currentState, destinationPlayerState)) {
            return
        }
        val elapsedTime = Util.elapsedTime
        videoTimeEnd = videoTime
        BitmovinLog.d(TAG, "Transitioning from $currentState to $destinationPlayerState")
        currentState.onExitState(this, elapsedTime, elapsedTime - elapsedTimeOnEnter, destinationPlayerState)
        elapsedTimeOnEnter = elapsedTime
        videoTimeStart = videoTimeEnd
        destinationPlayerState.onEnterState(this, data)
        currentState = destinationPlayerState
    }

    private fun isTransitionAllowed(
        currentState: PlayerState<*>?,
        destination: PlayerState<*>?,
    ): Boolean {
        if (destination === this.currentState) {
            return false
        } else if (this.currentState === PlayerStates.VIDEOSTARTFAILED) {
            return false
        } else if (currentState === PlayerStates.AD &&
            destination !== PlayerStates.ERROR && destination !== PlayerStates.ADFINISHED
        ) {
            return false
        } else if (currentState === PlayerStates.READY &&
            destination !== PlayerStates.ERROR && destination !== PlayerStates.STARTUP && destination !== PlayerStates.AD
        ) {
            return false
        } else if (currentState === PlayerStates.STARTUP && destination !== PlayerStates.READY &&
            destination !== PlayerStates.ERROR && destination !== PlayerStates.VIDEOSTARTFAILED &&
            destination !== PlayerStates.PLAYING && destination !== PlayerStates.AD
        ) {
            return false
        }
        return true
    }

    fun subscribe(listener: StateMachineListener) {
        listeners.subscribe(listener)
    }

    fun release() {
        listeners.clear()
        qualityChangeEventLimiter.release()
        bufferingTimeoutTimer.unsubscribe(::onRebufferingTimerFinished)
        videoStartTimeoutTimer.unsubscribe(::onVideoStartTimeoutTimerFinished)
    }

    fun addStartupTime(elapsedTime: Long) {
        startupTime += elapsedTime
    }

    fun getAndResetPlayerStartupTime() = analytics.getAndResetPlayerStartupTime()

    fun error(
        videoTime: Long,
        errorCode: ErrorCode,
    ) {
        transitionState(PlayerStates.ERROR, videoTime, errorCode)
    }

    fun sourceChange(
        oldVideoTime: Long,
        newVideoTime: Long,
        shouldStartup: Boolean,
    ) {
        transitionState(PlayerStates.SOURCE_CHANGED, oldVideoTime, null)
        resetSourceRelatedState()
        if (shouldStartup) {
            transitionState(PlayerStates.STARTUP, newVideoTime, null)
        }
    }

    fun pause(position: Long) {
        if (isStartupFinished) {
            transitionState(PlayerStates.PAUSE, position)
        } else {
            transitionState(PlayerStates.READY, position)
        }
    }

    fun startAd(position: Long) {
        transitionState(PlayerStates.AD, position)
        startupTime = 0
    }

    fun endAd() {
        // we don't need position here because during ads it would reflect videoTime of the ad and not
        // of the video being played
        transitionState(PlayerStates.ADFINISHED, videoTimeEnd)
    }

    fun closeCurrentSampleForCustomDataChangeIfNeeded(position: Long) {
        val originalState = currentState
        val shouldTransition = isPlayingOrPaused
        if (shouldTransition) {
            this.transitionState(PlayerStates.CUSTOMDATACHANGE, position)
            this.transitionState(originalState, position)
        }
    }

    fun seekStarted(videoTime: Long) {
        transitionState(PlayerStates.SEEKING, videoTime)
    }

    // This can be used as a template for all the quality changed methods,
    // as it is correctly setting the old values in the StateMachineListener
    fun subtitleChanged(
        videoTime: Long,
        oldValue: SubtitleDto?,
        newValue: SubtitleDto?,
    ) {
        if (!isStartupFinished) return
        if (!isPlayingOrPaused) return
        if (oldValue?.equals(newValue) == true) return
        val originalState = currentState
        transitionState(PlayerStates.SUBTITLECHANGE, videoTime, oldValue)
        transitionState(originalState, videoTime)
    }

    fun videoQualityChanged(
        videoTime: Long,
        didQualityChange: Boolean,
        setQualityFunction: () -> Unit,
    ) {
        qualityChanged(videoTime, didQualityChange, setQualityFunction)
    }

    fun audioQualityChanged(
        videoTime: Long,
        didQualityChange: Boolean,
        setQualityFunction: () -> Unit,
    ) {
        qualityChanged(videoTime, didQualityChange, setQualityFunction)
    }

    fun exitBeforeVideoStart(position: Long) {
        videoStartFailedReason = VideoStartFailedReason.PAGE_CLOSED
        transitionState(PlayerStates.VIDEOSTARTFAILED, position)
    }

    fun triggerSampleIfPlaying(ssaiRelated: Boolean = false) {
        if (currentState != PlayerStates.PLAYING) {
            return
        }

        triggerSample(ssaiRelated)
    }

    private fun qualityChanged(
        videoTime: Long,
        didQualityChange: Boolean,
        setQualityFunction: () -> Unit,
    ) {
        val originalState = currentState
        try {
            if (!isStartupFinished) return
            if (!qualityChangeEventLimiter.isQualityChangeEventEnabled) return
            if (!isPlayingOrPaused) return
            if (!didQualityChange) return
            transitionState(PlayerStates.QUALITYCHANGE, videoTime)
        } finally {
            // we always want to set the new quality, no matter if we transitioned states
            setQualityFunction()
        }
        transitionState(originalState, videoTime)
    }

    private val isPlayingOrPaused
        get() = currentState === PlayerStates.PLAYING || currentState === PlayerStates.PAUSE

    private fun onVideoStartTimeoutTimerFinished() {
        BitmovinLog.d(TAG, "VideoStartTimeout finish")
        videoStartFailedReason = VideoStartFailedReason.TIMEOUT
        transitionState(PlayerStates.VIDEOSTARTFAILED, 0, null)
    }

    private fun onRebufferingTimerFinished() {
        BitmovinLog.d(TAG, "rebufferingTimeout finish")

        error(
            playerContext.position,
            AnalyticsErrorCodes.ANALYTICS_BUFFERING_TIMEOUT_REACHED.errorCode,
        )
        disableRebufferHeartbeat()
        resetStateMachine()
    }

    object Factory {
        fun create(
            analytics: BitmovinAnalytics,
            playerContext: PlayerContext,
            looper: Looper,
            deviceInformationProvider: DeviceInformationProvider,
        ): PlayerStateMachine {
            val bufferingTimeoutTimer = ObservableTimer(Util.REBUFFERING_TIMEOUT.toLong(), 1000)
            val qualityChangeCountResetTimer =
                ObservableTimer(Util.ANALYTICS_QUALITY_CHANGE_COUNT_RESET_INTERVAL.toLong(), 1000)
            val qualityChangeEventLimiter = QualityChangeEventLimiter(qualityChangeCountResetTimer)
            val videoStartTimeoutTimer = ObservableTimer(Util.VIDEOSTART_TIMEOUT.toLong(), 1000)
            return PlayerStateMachine(
                analytics,
                bufferingTimeoutTimer,
                qualityChangeEventLimiter,
                videoStartTimeoutTimer,
                playerContext,
                looper,
                deviceInformationProvider,
            )
        }
    }

    companion object {
        private const val TAG = "PlayerStateMachine"
        private val rebufferingIntervals = arrayOf(3000, 5000, 10000, 30000, 59700)
    }
}
